parcels/particleset/baseparticleset.py

```diff
@@ -0,0 +1,38 @@
+from abc import ABC
+from abc import abstractmethod  # noqa
+
+
+class ParticleCollection(ABC):
+    """Interface."""
+
+
+class NDCluster(ABC):
+    """Interface."""
+
+
+class ParticleSetIterator:
+    """Interface?"""
+    def __init__(self, pset):
+        self.p = pset.data_accessor()
+        self.max_len = pset.size
+        self._index = 0
+
+    def __next__(self):
+        ''''Returns the next value from ParticleSet object's lists '''
+        if self._index < self.max_len:
+            self.p.set_index(self._index)
+            result = self.p
+            self._index += 1
+            return result
+        # End of Iteration
+        raise StopIteration
+
+
+class BaseParticleAccessor(ABC):
+    """Interface."""
```

Comment:
Well, related to my previous comment: miscommunication of ideas lead to this weird construct now where we try to derive abstract classes from specific classes. Hence, in my head, it would rather look like
```
class ParticleSetIterator(ABC):
    """Interface"""
    def ___init___(self):
        _head = None
        _tail = None
        _current = None

    def __next__(self):
        """... <fill purpose here> ...."""
        return None

    @property
    def head(self):
        return self._head

    @property
    def tail(self):
        return self._tail

    @property
    def current(self):
        return self._current

class BaseParticleAccessor(ParticleIterator):
    """Generic base class for accessing a Particle"""
    def __init__(self):
        super(BaseParticleAccessor, self).__init__()

    @property
    def current_cptr(self):
        p = self.current()
        if isinstance(p, Particle):
            return p.cptr()
        else:
            return -1

<in e.g. particlesets/trees/accessor.py>
from parcels.particlesets.baseparticleset import BaseParticleAccessor

class TreeParticleAccessor(BaseParticleAccessor):
    """Implementation of the ParticleAccessor for particles organized in a tree data structure"""
    def __init__(self):
        <fill in stuff here for what you need to parse a tree>

    def __next__(self):
        <parse your tree to fast-access the 'next' element>
        return p

    @property
    def current(self):
        <ideally your 'next' method stores some information that manipulates the '_current' variable>
        <perform some validity checks>
        return self._current
        
```

remember that in that latter implementation class, you do not need to override all functions - just the ones that needed to be left undefined in the base class as they depend on the underlying data structure. Also, ideally ALL functions in the base class make sense for ALL implementations - but if not so, the base class should per default return None, and is not overriden in the implementations.

Reply:
The current commit was only meant to show really the very basic outline of which files/classes we would end up using and how they should relate to each other. I deliberately did not implement any interfaces (i.e. abstract methods) yet, but we were mostly on the same page there. As discussed, I will merge the `ParticleSetIterator` interface into the `BaseParticleAccessor` abstract class, and not have it as a separate class.